cpp/src/neighbors/detail/vamana/macros.cuh

```diff
@@ -0,0 +1,92 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace cuvs::neighbors::vamana::detail {
+
+/* Macros to compute the shared memory requirements for CUB primitives used by search and prune */
+#define COMPUTE_SMEM_SIZES(degree, visited_size, DEG, CANDS)				\
+	if(degree == DEG && visited_size == CANDS) {					\
+	    search_smem_sort_size = static_cast<int>(sizeof(typename 			\
+	    cub::BlockMergeSort<DistPair<IdxT,accT>,32,CANDS/32>::TempStorage));	\
+											\
+	    prune_smem_sort_size = static_cast<int>(sizeof(typename 			\
+	    cub::BlockMergeSort<DistPair<IdxT,accT>,32,(CANDS+DEG)/32>::TempStorage));	\
+	}										
+	
+// Current supported sizes for degree and visited_size. Note that visited_size must be > degree
+#define SELECT_SMEM_SIZES(degree, visited_size)				\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 32, 64);		\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 32, 128);		\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 32, 256);		\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 32, 512);		\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 64, 128);		\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 64, 256);		\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 64, 512);		\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 128, 256);		\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 128, 512);		\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 256, 512);		\
+	COMPUTE_SMEM_SIZES(degree, visited_size, 256, 1024);		
+
+
+
+/* Macros to call the CUB BlockSort primitives for supported sizes for ROBUST_PRUNE*/
+#define PRUNE_CALL_SORT(degree, visited_list, DEG, CANDS)			\
+	if(degree == DEG && visited_list == CANDS) {				\
+  		using BlockSortT = 						\
+		cub::BlockMergeSort<DistPair<IdxT,accT>,32,(DEG+CANDS)/32>;	\
+  		auto& sort_mem = 						\
+		reinterpret_cast<typename BlockSortT::TempStorage&>(smem);	\
+  		sort_edges_and_cands<accT,IdxT,DEG,CANDS>			\
+			(new_nbh_list, &query_list[i], &sort_mem);		\
+	}
+
+#define PRUNE_SELECT_SORT(degree, visited_list)                 	 \
+	PRUNE_CALL_SORT(degree, visited_size, 32, 64);			\
+	PRUNE_CALL_SORT(degree, visited_size, 32, 128);			\
+	PRUNE_CALL_SORT(degree, visited_size, 32, 256);			\
+	PRUNE_CALL_SORT(degree, visited_size, 32, 512);			\
+	PRUNE_CALL_SORT(degree, visited_size, 64, 128);			\
+	PRUNE_CALL_SORT(degree, visited_size, 64, 256);			\
+	PRUNE_CALL_SORT(degree, visited_size, 64, 512);			\
+	PRUNE_CALL_SORT(degree, visited_size, 128, 256);		\
+	PRUNE_CALL_SORT(degree, visited_size, 128, 512);		\
+	PRUNE_CALL_SORT(degree, visited_size, 256, 512);		\
+	PRUNE_CALL_SORT(degree, visited_size, 256, 1024);		
+
+
+
+/* Macros to call the CUB BlockSort primitives for supported sizes for GREEDY SEARCH */
+#define SEARCH_CALL_SORT(topk, CANDS)						\
+	if(topk == CANDS) {							\
```

Comment:
Would involve re-working the macros a bit for this, but agreed it would be better to accept non-power-of-2 and round the template parameter up to the next value. Will try to re-work this but need to re-run some benchmarks to verify no perf regression.

Reply:
Add support for non-power-of-2 values for visited_list. Gives warning that it is rounding up.